### PR TITLE
Feature/tao 9338/password grant type

### PR DESCRIPTION
**Pull request by tikhanovichA**
https://oat-sa.atlassian.net/browse/TAO-9338

For testing, you can use the next config for webhooks:

webhookPayloadFactory.conf.php:
```
return new oat\tao\model\webhooks\WebhookFileRegistry(array(
    'webhooks' => array(
        'testwh' => array(
            'id' => 'testwh',
            'url' => 'https://nccer--testing.my.salesforce.com/services/data/v46.0/sobjects/Test_Taken__c',
            'httpMethod' => 'GET',
            'retryMax' => 5,
            'auth' => array(
                'authClass' => '\\oat\\taoOauth\\model\\bootstrap\\OAuth2AuthType',
                'credentials' => array(
                    'grant_type' => 'password',
                    'client_id' => '3MVG9AzPSkglhtpvez2DfkLfoXrDEq4KyTfT0qpNitc3YsEFLTbcx2pJMu23otOXsOys6novR5A45FEwquGxW',
                    'client_secret' => '47100AAB48C40077150484CFC75FA11659E62274B54BACC7E2B0AC9CD02C573D',
                    'username' => 'tao.int@nccer.org.test',
                    'password' => 'webhook1ntegrationcKQds3TQPlLbjmAdRNLfb5eW',
                    'token_url' => 'https://test.salesforce.com/services/oauth2/token'
                )
            )
        )
    ),
    'events' => array(
        'oat\\taoDelivery\\models\\classes\\execution\\event\\DeliveryExecutionCreated' => ['testwh']
    )
));
```

### CODE REVIEW

File: model/OAuthClient.php [hunk 1]
```diff
@@ -207,7 +207,8 @@ protected function getResponse(RequestInterface $request, $options = [])
     protected function requestAccessToken($params = [])
     {
         $params = $this->addTokenRequestParameters($params);
-        $accessToken = $this->getProvider()->getAccessToken(self::DEFAULT_GRANT_TYPE, $params);
+        $grantType = !empty($this->getOption('grant_type')) ? $this->getOption('grant_type') : self::DEFAULT_GRANT_TYPE;
```
Comment (bartlomiejmarszal) [inline model/OAuthClient.php]: This change should be tested. This means that you should modify `OAuthClientTest`. you could make some refactoring in this class maybe?

File: model/storage/OauthCredentialsFactory.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * @package oat\taoOauth\model\storage
+ */
+class OauthCredentialsFactory extends ConfigurableService
+{
+    /**
+     * @param array $parameters
+     * @return OauthCredentials|ClientCredentialsType|PasswordType|AuthorizationCodeType
+     * @throws common_exception_ValidationFailed
+     */
+    public function getCredentialTypeByCredentials($parameters = [])
```
Comment (bartlomiejmarszal) [inline model/storage/OauthCredentialsFactory.php]: This have to be unit tested

Comment (tikhanovichA) [reply model/storage/OauthCredentialsFactory.php]: it's just mapping, why we need to test it?

File: model/OAuthClient.php (hunk 1 above)
Comment (tikhanovichA) [reply model/OAuthClient.php]: Fixed. https://github.com/oat-sa/extension-tao-oauth/pull/42/files#diff-ac10a0bf21843ed85cdb4aeeba4f2590R156

File: model/bootstrap/OAuth2AuthType.php [hunk 3]
```diff
@@ -117,21 +115,31 @@ public function getTemplate()
...
     {
+        /** @noinspection PhpIncompatibleReturnTypeInspection */
         return $this->getServiceLocator()->get(Oauth2Service::SERVICE_ID);
     }
 
     /**
-     * @return \core_kernel_classes_Resource|void
+     * @return OauthCredentialsFactory
+     */
+    protected function getOauthCredentialsFactory()
```
Comment (alexeyvy) [inline model/bootstrap/OAuth2AuthType.php]:
```suggestion
    private function getOauthCredentialsFactory()
```

File: test/model/Oauth2TypeTest.php [hunk 4]
```diff
@@ -4,110 +4,61 @@
...
-    public function testOAuth2Type()
+    /**
+     * @dataProvider getGrantTypesCredentials
+     *
+     * @param $dataProvider
+     * @throws \ConfigurationException
+     * @throws \common_Exception
+     * @throws \oat\taoOauth\model\exception\OauthException
+     */
+    public function testOauth2TypeWithDifferentsGrantTypes($dataProvider)
```
Comment (alexeyvy) [inline test/model/Oauth2TypeTest.php]:
```suggestion
    public function testOauth2TypeWithDifferentGrantTypes($dataProvider)
```

File: test/model/Oauth2TypeTest.php [hunk 5]
```diff
@@ -4,110 +4,61 @@
...
-                'headers' => null
-            ]);
-
-        $Oauth2ServiceMock->method('getClient')->willReturn($clientMock);
-
-        $serviceLocatorMock = $this->getServiceLocatorMock([
-            Oauth2Service::SERVICE_ID => $Oauth2ServiceMock
-        ]);
-
+        $Oauth2ServiceMock = $this->getOauth2ServiceMock($dataProvider['out']);
```
Comment (alexeyvy) [inline test/model/Oauth2TypeTest.php]:
```suggestion
        $oauth2ServiceMock = $this->getOauth2ServiceMock($dataProvider['out']);
```

File: test/model/Oauth2TypeTest.php [hunk 6]
```diff
@@ -4,110 +4,61 @@
...
         $authType = new OAuth2AuthType;
 
-        $authType->setServiceLocator($serviceLocatorMock);
-
-        $authType->setCredentials($this->credentials);
+        $authType->setServiceLocator($this->getServiceLocator([Oauth2Service::SERVICE_ID => $Oauth2ServiceMock]));
```
Comment (alexeyvy) [inline test/model/Oauth2TypeTest.php]:
```suggestion
        $authType->setServiceLocator($this->getServiceLocator([Oauth2Service::SERVICE_ID => $oauth2ServiceMock]));
```

File: model/storage/grant/AuthorizationCodeType.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+use oat\tao\model\auth\AbstractCredentials;
+use oat\taoOauth\model\provider\Provider;
+
+/**
+ * Class AuthorizationCodeType
+ * @package oat\taoOauth\model\storage\grant
+ */
+class AuthorizationCodeType extends OauthCredentials
+{
+    const NAME = 'authorization_code';
```
Comment (alexeyvy) [inline model/storage/grant/AuthorizationCodeType.php]:
```suggestion
    const NAME = 'authorization_code';

```

File: model/storage/OauthCredentialsFactory.php (hunk 2 above)
Comment (alexeyvy) [inline model/storage/OauthCredentialsFactory.php]:
```suggestion
    public function getCredentialTypeByCredentials(array $parameters = [])
```

File: model/storage/OauthCredentialsFactory.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Class OauthCredentials
+ * @package oat\taoOauth\model\storage
+ */
+class OauthCredentialsFactory extends ConfigurableService
+{
+    /**
+     * @param array $parameters
+     * @return OauthCredentials|ClientCredentialsType|PasswordType|AuthorizationCodeType
```
Comment (alexeyvy) [inline model/storage/OauthCredentialsFactory.php]:
```suggestion
     * @return OauthCredentials
```

File: model/OAuthClient.php [hunk 9]
```diff
@@ -233,17 +234,18 @@ protected function addTokenRequestParameters($params = [])
...
      *  Get stored access token. If there is no token in the storage or token has expired then request new token.
      *
+     * @param array $options
      * @return AccessToken
      * @throws OauthException
      * @throws \ConfigurationException
      * @throws \common_Exception
      */
-    protected function getAccessToken()
+    protected function getAccessToken($options = [])
```
[the remark this answers is not in the document]
Comment (alexeyvy) [reply model/OAuthClient.php]: not sure, may it be private?

File: model/storage/OauthCredentialsFactory.php (hunk 2 above)
Comment (tikhanovichA) [reply model/storage/OauthCredentialsFactory.php]: Done

File: test/model/OauthCredentialsFactoryTest.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Class OauthCredentialsFactoryTest
+ */
+class OauthCredentialsFactoryTest extends TestCase
+{
+    /**
+     * @throws \common_exception_ValidationFailed
+     */
+    public function testGettingCredentialTypeByCredentials()
```
Comment (bartlomiejmarszal) [inline test/model/OauthCredentialsFactoryTest.php]: Maybe this doesn't matter here but you should have a separate test method for each expected call from this method. You never know if for a newly created object this method will not behave differently. You also need to perform expect throw error assertion.

Comment (tikhanovichA) [reply test/model/OauthCredentialsFactoryTest.php]: got it. My bad

File: test/model/OauthCredentialsFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        $this->assertInstanceOf(PasswordType::class, $passwordGrantType);
+    }
+
+    /**
+     * @throws common_exception_ValidationFailed
+     */
+    public function testGettingAuthCodeTypeByCredentials()
+    {
+        $OauthCredentialsFactory = new OauthCredentialsFactory();
+
```
Comment (bartlomiejmarszal) [inline test/model/OauthCredentialsFactoryTest.php]:
```suggestion
/* @noinspection PhpUnhandledExceptionInspection */
```

File: test/model/OauthCredentialsFactoryTest.php [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+ */
+class OauthCredentialsFactoryTest extends TestCase
+{
+    /**
+     * @throws common_exception_ValidationFailed
+     */
+    public function testGettingClientCredentialTypeByCredentials()
+    {
+        $OauthCredentialsFactory = new OauthCredentialsFactory();
+
```
Comment (bartlomiejmarszal) [inline test/model/OauthCredentialsFactoryTest.php]: /* @noinspection PhpUnhandledExceptionInspection */

**Comment by tikhanovichA** [comment]
> Can we remove phpdocs and paste `@noinspection`

fixed

File: model/storage/OauthCredentialsFactory.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+            return new PasswordType($parameters);
+        }
+        if ($grantType === ClientCredentialsType::NAME) {
+            return new ClientCredentialsType($parameters);
+        }
+        if ($grantType === AuthorizationCodeType::NAME) {
+            return new AuthorizationCodeType($parameters);
+        }
+
+        throw new common_exception_ValidationFailed($parameters[Provider::GRANT_TYPE]);
```
Comment (siarhei-baradzin) [inline model/storage/OauthCredentialsFactory.php]: `common_exception_ValidationFailed` expects not field value but field name as an argument

Comment (tikhanovichA) [reply model/storage/OauthCredentialsFactory.php]: I will create a PR for that in generis

**Pull request merged**